### ISSUE TITLE
Kippler support

### ISSUE THREAD

**Issue by hackebike** [markniu/Bed_Distance_sensor#1]
Any chance you add Kippler support?

**Comment by markniu** [comment]
Seems will wait for more time, Since I am not familiar with klipper and python, also try to looking someone help to supporting.

**Comment by HelgeKeck** [comment]
hi there,

working on klipper support for the sensor but cant even connect the probe for any reason. Its connected to the raspebrry PI`s I2C connector but i2cdetect cant find the deivce.

any idea?

**Comment by HelgeKeck** [comment]
The device has a 5V VIN, but does the device uses a 3.3V or a 5V I2C connection. Cant find any information on it

**Comment by HelgeKeck** [comment]
I tried it with an Arduino I2C scanner nd found this

if the device is connected to the arduino it returns an derror for these addresses

Unknown error at address 0x51
Unknown error at address 0x54
Unknown error at address 0x57
Unknown error at address 0x5A
Unknown error at address 0x5D

**Comment by markniu** [comment]
there is a different from the standard I2C.  See this timing diagram of this sensor  https://github.com/markniu/Bed_Distance_sensor/blob/main/doc/images/0220517153950.png , for increasing the speed, there is no I2C address transfer in the timing diagram that's the difference from the standard I2C.
There is a library for the arduino:  markyue/Panda_SoftMasterI2C   
for example add 
```
lib_deps          = 
  markyue/Panda_SoftMasterI2C
```
in the platformio.ini

**Comment by HelgeKeck** [comment]
ok, thank you

**Comment by markniu** [comment]
> The device has a 5V VIN, but does the device uses a 3.3V or a 5V I2C connection. Cant find any information on it

this device is 5v I2C connection,but can work with normal 3.3v MCU like stm32 which most gpios are 5V tolerate

**Comment by HelgeKeck** [comment]
nice it works, getting the first readings. will play a bit with it and let you knwo when i need more infos

**Comment by markniu** [comment]
> nice it works, getting the first readings. will play a bit with it and let you know when i need more infos

look forward to

**Comment by HelgeKeck** [comment]
https://www.youtube.com/watch?v=nbHTrJHoE-k

**Comment by markniu** [comment]
wow, nice work!

**Comment by HelgeKeck** [comment]
Thats not good. I just realized that this sensor will most likely not be supported by klipper naytime soon. Problem is the unusual I2C communication the sensor uses. We can use the inbuilt I2C micricontroller code from klipper to connect to the device, a microcontroller implementation just for this snesor seems unlikely to be approved by the klipper mantainers. So a real implementation is not likely.

**Comment by HelgeKeck** [comment]
is it possible to change the firmware of the sensor to use standarized I2C communication?

**Comment by markniu** [comment]
it can not be flashed firmware from the i2c port.
and the reason for not use standard I2C is that we need quik response of the sensor when homing z axis.
maybe we can add the softI2C code of this sensor to the klipper.

**Comment by HelgeKeck** [comment]
i am trying to implement it atm

**Comment by HelgeKeck** [comment]
![image](https://user-images.githubusercontent.com/13624572/200111333-21cf988c-0a92-457d-bd3f-a868a196bb46.png)

**Comment by markniu** [comment]
> ![image](https://user-images.githubusercontent.com/13624572/200111333-21cf988c-0a92-457d-bd3f-a868a196bb46.png)

Can BD sensor data be read with klipper now?

**Comment by HelgeKeck** [comment]
yes, on a microcontroller level

**Comment by markniu** [comment]
good job ! I am editing the wiki document about the calibration that maybe you needed

**Comment by HelgeKeck** [comment]
yes, working on the calibration, having also some questions. will ping oyu in the nbext days

**Comment by markniu** [comment]
Have updated the document which contains the calibration process.
https://github.com/markniu/Bed_Distance_sensor/wiki/How-it-works

**Comment by ashbuilds** [comment]
Hi @HelgeKeck @markniu, may I know what are the configs for Klipper implementation?

**Comment by JayBee-git** [comment]
I was wondering, could we simply use a MCU to connect the sensor to klipper, like can do with the ADXL345 sensor (for resonance compensation)? Would it solve the issue with the unusual I2C protocol, or would it be too slow to react?

**Comment by ashbuilds** [comment]
@JayBee-git I was also thinking the same but not sure how Klipper will read the data from sensor. I'm also looking into it.

**Comment by JayBee-git** [comment]
@ashbuilds true, I know serial can be used in klipper but that would probably require to change some code. BLtouch is going through the printer motherboard which is a MCU as well, so I bet there is something that can be done here. Maybe even connecting the BDTouch to the motherboard could work?

**Comment by bombela** [comment]
The output of the TCA 305 sensor chip is a simple open drain already. What is the point of using a custom protocol with a micro controller here instead of simply exposing the output of the sensor directly? I am sure you have good reasons @markniu but it would be nice if you could explain them.

**Comment by markniu** [comment]
There is a progress for Bed Distance sensor working with klipper:
can display the distance value on screen, homing, and still need some work for the probe, calibration ...
![image](https://user-images.githubusercontent.com/13879340/209631123-f7c1b604-1370-4c83-998c-8f8571555a51.png)

the BDsensor is connected to mainboard not RaspberryPi, It's not recommend to connect to RaspberryPi for there maybe a sync problem while homing.

**Comment by Townie-au** [comment]
Awesome!
Really very interested in this.
Does it work with a pei bed surface being magnetic?
I have glass ATM. But thinking about getting pei.

**Comment by markniu** [comment]
> Awesome! Really very interested in this. Does it work with a pei bed surface being magnetic? I have glass ATM. But thinking about getting pei.

yes, I also test with the pei + steel + magnet on the bed

**Comment by neshami** [comment]
> There is a progress for Bed Distance sensor working with klipper:
> can display the distance value on screen, homing, and still need some work for the probe, calibration ...
> ![image](https://user-images.githubusercontent.com/13879340/209631123-f7c1b604-1370-4c83-998c-8f8571555a51.png)
> 
> the BDsensor is connected to mainboard not RaspberryPi, It's not recommend to connect to RaspberryPi for there maybe a sync problem while homing.
> 

Have you any news about bd sensor on klipper?

**Comment by markniu** [comment]
> 
sorry for that delay maybe still need about 2 weeks for the support klipper software, since most of my family got the covid19 last 2 weeks and all recovery fine now. I will have more free time this month from now on.

**Comment by markniu** [comment]
Here is the beta version of Klipper that support Bed Distance Sensor
https://github.com/markniu/Bed_Distance_sensor/wiki/Installing-for-Klipper

**Comment by JayBee-git** [comment]
That's really cool, thanks a ton for your work!
I'll install that on another SD card and report back.

**Comment by Townie-au** [comment]
> Here is the beta version of Klipper that support Bed Distance Sensor https://github.com/markniu/Bed_Distance_sensor/wiki/Installing-for-Klipper

Is it possible to use with Mainsail?

**Comment by cubiq** [comment]
is there any chance for this to be included/accepted into the official klipper repo?

**Comment by markniu** [comment]
> Mainsail

I am not use the Mainsail before, but I think it's no problem to work with mainsail for that all the changed source code for BDsensor are in klipper side.

**Comment by markniu** [comment]
> is there any chance for this to be included/accepted into the official klipper repo?

I will commit the changed source code into official klipper github after doing more test

**Comment by markniu** [comment]
The code has been pulled to klipper github: https://github.com/Klipper3d/klipper/pull/6031

**Comment by cubiq** [comment]
waiting for it to be approved to buy a few sensors.

PS: I believe you need to sign-off with your full real name (no alias)

**Comment by ashbuilds** [comment]
That's awesome! I will try it out asap, thanks!

**Comment by henryrgithub** [comment]
@markniu I'm interested in this sensor as I find my current solution inadequate, but am not really interested in keeping a non-mainline firmware installed on my printer long-term. Do you still plan to try and submit a PR to Klipper master at some point? I don't see any commits to your klipper fork since 5/15. I am familiar with Python and C, so if there's anything specific that needs refinement on the fork, if you create issues for it I can take a look.

**Comment by benbender** [comment]
@markniu @henryrgithub There is a new PR at https://github.com/Klipper3d/klipper/pull/6490#issuecomment-2014302923 that needs to incorporate some requested changes. From my feeling, if those are done, the chances would be good to get it finally into mainline!

**Comment by henryrgithub** [comment]
The main thing I see glancing over it is the effort to use a non-sensor-specific probe functions, or at least just have the functions shared with other current eddy sensors. Seems like a reasonable lift, I'll take a look at it this coming week.
